internal/pkg/workspace/workspace.go

```diff
@@ -236,31 +241,42 @@ func (ws *Workspace) ListAddonFiles(appName string) ([]string, error) {
 	return addonsFileNames, nil
 }
 
-// ReadAddonFiles reads all addon files.
+// ReadAddonFiles reads addon files for an application.
 func (ws *Workspace) ReadAddonFiles(appName string) (*ReadAddonFilesOutput, error) {
-	addonFiles, err := ws.ListAddonFiles(appName)
+	addonFileNames, err := ws.ListAddonFiles(appName)
 	if err != nil {
-		return nil, fmt.Errorf("list addon files: %w", err)
+		return nil, err
 	}
-	var resources, params, outputs string
-	for _, fileName := range addonFiles {
+	addonFiles := map[string]string{paramsFile: "", outputsFile: "", resourcesFiles: ""}
+	for _, fileName := range addonFileNames {
 		content, err := ws.read(appName, addonsDirName, fileName)
+		trimmedContent := strings.TrimSpace(string(content))
 		if err != nil {
 			return nil, fmt.Errorf("read addon file %s: %w", fileName, err)
 		}
 		switch fileName {
-		case paramsFileName:
-			params = string(content)
-		case outputsFileName:
-			outputs = string(content)
+		case paramsFile:
+			addonFiles[paramsFile] = trimmedContent
+		case outputsFile:
+			addonFiles[outputsFile] = trimmedContent
 		default:
-			resources += string(content)
+			addonFiles[resourcesFiles] += trimmedContent + "\n"
 		}
 	}
+	var missingFiles []string
+	for fileName, content := range addonFiles {
+		if content == "" {
+			missingFiles = append(missingFiles, fileName)
+		}
+	}
+	sort.Strings(missingFiles)
+	if len(missingFiles) != 0 {
+		return nil, fmt.Errorf("addons directory missing file(s): %s", strings.Join(missingFiles, ", "))
```

Comment:
This is very cool! My only worry is that seeing `"{resources}.yml"` in the error message wont be clear enough to a user.

Is it possible to format the error like this:
`addons directory is missing required files(s): "params.yml", "outputs.yml", and at least one resource YAML file such as "s3-bucket.yml". `

Reply:
Good callout!